File: backend/app/schemas/ai.py
from pydantic import BaseModel, Field, field_validator, model_validator


class NoteSummaryResponse(BaseModel):
    note_id: int
    summary: str


class MaterialQuestionRequest(BaseModel):
    question: str = Field(min_length=1, max_length=2000)

    @field_validator("question")
    @classmethod
    def question_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Question must not be blank.")
        return value.strip()


class MaterialQuestionResponse(BaseModel):
    material_id: int
    answer: str


class StudyPlanRequest(BaseModel):
    subject_ids: list[int] = Field(default_factory=list, max_length=20)
    days: int = Field(default=7, ge=1, le=30)
    minutes_per_day: int = Field(default=60, ge=15, le=480)
    priorities: str | None = Field(default=None, max_length=2000)

    @field_validator("subject_ids")
    @classmethod
    def subject_ids_must_be_positive(cls, value: list[int]) -> list[int]:
        if any(subject_id <= 0 for subject_id in value):
            raise ValueError("Subject IDs must be positive.")
        if len(set(value)) != len(value):
            raise ValueError("Subject IDs must be unique.")
        return value

    @field_validator("priorities")
    @classmethod
    def priorities_must_not_be_blank(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("Priorities must not be blank.")
        return value.strip() if value is not None else value


class StudyPlanResponse(BaseModel):
    plan: str


class QuizGenerationRequest(BaseModel):
    subject_id: int | None = Field(default=None, gt=0)
    material_id: int | None = Field(default=None, gt=0)
    question_count: int = Field(default=5, ge=1, le=20)
    topic: str | None = Field(default=None, max_length=500)

    @field_validator("topic")
    @classmethod
    def topic_must_not_be_blank(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("Topic must not be blank.")
        return value.strip() if value is not None else value

    @model_validator(mode="after")
    def require_single_source(self):
        if (self.subject_id is None) == (self.material_id is None):
            raise ValueError("Provide exactly one subject_id or material_id.")
        return self


class PracticeQuestionRequest(BaseModel):
    subject_id: int | None = Field(default=None, gt=0)
    material_id: int | None = Field(default=None, gt=0)
    topic: str | None = Field(default=None, max_length=500)

    @field_validator("topic")
    @classmethod
    def topic_must_not_be_blank(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("Topic must not be blank.")
        return value.strip() if value is not None else value

    @model_validator(mode="after")
    def require_single_source(self):
        if (self.subject_id is None) == (self.material_id is None):
            raise ValueError("Provide exactly one subject_id or material_id.")
        return self


class GeneratedQuizQuestion(BaseModel):
    question: str = Field(min_length=1, max_length=2000)
    options: list[str] = Field(min_length=2, max_length=6)
    correct_answer: str = Field(min_length=1, max_length=500)
    explanation: str = Field(min_length=1, max_length=2000)

    @field_validator("question", "correct_answer", "explanation")
    @classmethod
    def generated_text_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Generated quiz text must not be blank.")
        return value.strip()

    @field_validator("options")
    @classmethod
    def options_must_be_valid(cls, value: list[str]) -> list[str]:
        normalized = [option.strip() for option in value]
        if any(not option for option in normalized):
            raise ValueError("Quiz options must not be blank.")
        if len(set(normalized)) != len(normalized):
            raise ValueError("Quiz options must be unique.")
        return normalized

    @model_validator(mode="after")
    def correct_answer_must_be_an_option(self):
        if self.correct_answer not in self.options:
            raise ValueError("The correct answer must match one option.")
        return self


class GeneratedQuizResponse(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    questions: list[GeneratedQuizQuestion] = Field(min_length=1, max_length=20)

    @field_validator("title")
    @classmethod
    def title_must_not_be_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("Generated quiz title must not be blank.")
        return value.strip()

File: mobile/src/types/ai.ts
// ─── Nexora Mobile — AI Type Definitions ────────────────────────────────────
// Mirrors backend/app/schemas/ai.py exactly.
// Import from '@/types/ai' in services and components.

// ─── Note Summarization ───────────────────────────────────────────────────────
/** Response from POST /notes/{note_id}/summarize */
export interface NoteSummaryResponse {
  note_id: number;
  summary: string;
}

// ─── Study-Material Q&A ───────────────────────────────────────────────────────
/** Request body for POST /study-materials/{material_id}/ask */
export interface MaterialQuestionRequest {
  question: string;
}

/** Response from POST /study-materials/{material_id}/ask */
export interface MaterialQuestionResponse {
  material_id: number;
  answer: string;
}

// ─── AI Study-Plan Generation ─────────────────────────────────────────────────
/** Request body for POST /study-plans/generate
 *  - subject_ids: [] means all subjects
 *  - days: 1–30 (default 7)
 *  - minutes_per_day: 15–480 (default 60)
 *  - priorities: optional free-text up to 2000 chars
 */
export interface StudyPlanRequest {
  subject_ids?: number[];
  days?: number;
  minutes_per_day?: number;
  priorities?: string;
}

/** Response from POST /study-plans/generate */
export interface StudyPlanResponse {
  id?: number;
  plan: string;
  title?: string;
  subject_ids?: number[];
  days?: number;
  minutes_per_day?: number;
  priorities?: string;
}

export interface AIConversation {
  id: number;
  user_id: number;
  title: string;
  created_at: string;
  updated_at: string;
}

export interface AIMessage {
  id: number;
  conversation_id: number;
  role: 'user' | 'assistant';
  content: string;
  created_at: string;
}

// ─── AI Quiz Generation ───────────────────────────────────────────────────────
/** Request body for POST /quizzes/generate
 *  Exactly one of subject_id or material_id must be provided.
 *  - question_count: 1–20 (default 5)
 *  - topic: optional hint up to 500 chars
 */
export interface QuizGenerationRequest {
  subject_id?: number;
  material_id?: number;
  question_count?: number;
  topic?: string;
}

/** A single generated quiz question (not persisted) */
export interface GeneratedQuizQuestion {
  question: string;
  options: string[];
  correct_answer: string;
  explanation: string;
}

export type PracticeQuestionResponse = GeneratedQuizQuestion;

export interface QuizExplanationResponse {
  quiz_id: number;
  question_id: number;
  explanation: string;
}

/** Response from POST /quizzes/generate (ephemeral — not saved to DB) */
export interface GeneratedQuizResponse {
  title: string;
  questions: GeneratedQuizQuestion[];
}

// ─── Shared AI Error Type ─────────────────────────────────────────────────────
/**
 * Normalised error kind returned by parseAIError().
 *
 * rate_limit  — HTTP 429, rolling 24-hour quota exhausted
 * validation  — HTTP 422, invalid request payload
 * not_found   — HTTP 404, resource doesn't exist / file missing
 * server      — HTTP 502/503, upstream AI error or not configured
 * network     — timeout or no response
 * unknown     — anything else
 */
export type AIErrorKind =
  | 'rate_limit'
  | 'validation'
  | 'not_found'
  | 'server'
  | 'network'
  | 'unknown';
